File: src/MyServiceBus.Http/HttpSendTransport.cs
namespace MyServiceBus;

public sealed class HttpSendTransport : ISendTransport
{
    private readonly HttpClient _client;
    private readonly Uri _address;

    public HttpSendTransport(HttpClient client, Uri address)
    {
        _client = client;
        _address = address;
    }

    [Throws(typeof(HttpRequestException), typeof(TaskCanceledException), typeof(UriFormatException))]
    public async Task Send<T>(T message, SendContext context, CancellationToken cancellationToken = default)
        where T : class
    {
        var body = await context.Serialize(message);
        var request = new HttpRequestMessage(HttpMethod.Post, _address);
        request.Content = new ByteArrayContent(body.ToArray());

        if (context.Headers != null)
        {
            foreach (var header in context.Headers)
            {
                request.Headers.TryAddWithoutValidation(header.Key, header.Value?.ToString());
            }
        }

        if (request.Content.Headers.ContentType == null)
            request.Content.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue("application/json");

        var response = await _client.SendAsync(request, cancellationToken);
        response.EnsureSuccessStatusCode();
    }
}

File: src/Java/myservicebus-http/src/main/java/com/myservicebus/http/HttpSendTransport.java
package com.myservicebus.http;

import com.myservicebus.SendTransport;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Map;

public class HttpSendTransport implements SendTransport {
    private final HttpClient client;
    private final URI address;

    public HttpSendTransport(HttpClient client, URI address) {
        this.client = client;
        this.address = address;
    }

    @Override
    public void send(byte[] data, Map<String, Object> headers, String contentType) {
        try {
            HttpRequest.Builder builder = HttpRequest.newBuilder(address)
                    .POST(HttpRequest.BodyPublishers.ofByteArray(data));
            if (contentType != null) {
                builder.header("Content-Type", contentType);
            }
            if (headers != null) {
                for (Map.Entry<String, Object> entry : headers.entrySet()) {
                    if (entry.getValue() != null) {
                        builder.header(entry.getKey(), entry.getValue().toString());
                    }
                }
            }
            client.send(builder.build(), HttpResponse.BodyHandlers.discarding());
        } catch (Exception ex) {
            throw new RuntimeException("Failed to send HTTP request", ex);
        }
    }
}
